### ISSUE TITLE
Cannot ignore passwords from keepassxc

### ISSUE THREAD

**Issue by bt103** [hluk/CopyQ#1744]
**Describe the bug**
I tried using this [link](https://copyq.readthedocs.io/en/latest/faq.html#how-to-omit-storing-text-copied-from-specific-windows-like-a-password-manager) to setup copyq to ignore passwords copied to the clipboard from keepassxc.
_When I copy a password from keepassxc, copyq will ignore the first password and all subsequent passwords that are different than the previous password.  However as soon as I copy the same password twice in a row, it shows up in copyq._

**--UPDATE - it seems the behaviour where the password shows-up in copyq is totally random and not only when a password is copied twice in a row.  I had a good run of about a dozen times where the password was ignored, then it started happening a few times.  Then I had a solid sting of about a half dozen times where copyq would remember the password every time.  Now i'm back on a run where it is working properly.  I make no changes during this testing.
I've italicised original text that is incorrect (I cannot find a strikethrough format).**

**To Reproduce**
Setup copyq as per this [link](https://copyq.readthedocs.io/en/latest/faq.html#how-to-omit-storing-text-copied-from-specific-windows-like-a-password-manager).
Create at least two passwords in keepassxc.
In keepassxc, copy passwords to the clipboard using the button in the top menu.
_Copy the same password to the clipboard twice in a row and notice the password shows-up in copyq.
Note that I was waiting more than 10 seconds between each copy which is the time it takes for keepassxc to clear the clipboard automatically._

--See update note above where the problem happens randomly.

**Expected behavior**
The rule should prevent passwords from showing-up in copyq,
_Interestingly, the clipboard is empty before every copy because keepassxc cleared the clipboard on the previous copy.  So it is odd that "something" must have a history of the clipboard if it does the unusual behaviour when the same password is copied twice in a  row._

**Screenshots**
n/a

**Version, OS and Environment**

- Linux Mint19 xfce
- copyq v4.1.0~bionic
- keepassxc 2.6.6

**Additional context**
If it helps, when it is working properly and copyq ignores the password, the title bar of copyq will only display "Copyq" for the 10 seconds that the item is in the clipboard and then will display "Empty - Copyq".

When it is behaving badly, the title bar of copyq will display the "_password_ - Copyq" for the 10 seconds that the item is in the clipboard, and then will display "Empty - Copyq".  Of course, by that time, the password shows-up in the history list.

**Comment by hluk** [comment]
I'm using keepassxc 2.6.6 with CopyQ without any issue (passwords are not stored in CopyQ). I'm using following command:

```ini
[Command]
Automatic=true
Command="
    copyq ignore"
Icon=\xf069
Name=Ignore *\"Password\"* window
Remove=true
Window=" - KeePass|Generate Password"
```

Can you share logs (F12 shortcut from main window)?

Does the keepassxc window hide or loses focus right after the password is copied?

Are you using X11 or Wayland?

**Comment by bt103** [comment]
> I'm using keepassxc 2.6.6 with CopyQ without any issue (passwords are not stored in CopyQ). I'm using following command:
> 
> ```ini
> [Command]
> Automatic=true
> Command="
>     copyq ignore"
> Icon=\xf069
> Name=Ignore *\"Password\"* window
> Remove=true
> Window=" - KeePass|Generate Password"
> ```
> 
> Can you share logs (F12 shortcut from main window)?
> 
> Does the keepassxc window hide or loses focus right after the password is copied?
> 
> Are you using X11 or Wayland? **Edit:** Thought since it's XFCE, I'm assuming it is X11.

Thanks for looking at this.  Here is the command text:
```
[Command]
Automatic=true
Command="
    copyq ignore"
Icon=\xf069
Name=Ignore *\"Password\"* window
Remove=true
Window=KeePass
```
Keepass is setup to not hide when copy  button is pressed.
I'm using X11.
The f12 log screen doesn't show any activity for today (perhaps I need to adjust a log level).

However,  I have had a breakthrough and can now consistently reproduce the problem (forgive me as I thought I had recognised the pattern in the past, but this time I'm pretty sure).  To reproduce, first go into copyq window and perform an "activity".  I know of two activities that will work (but maybe there are others).  One activity is to delete the last entry using the trash can button.  The other activity is to open the "Command/Global Shortcuts" dialogue and close it with the OK button.  Then go to keepassxc and copy a password.  Bingo, shows up in copyq every time.

I hope that explanation will suffice.  If not, I could look for a screen capture program.
So basically go to copyq first and delete the latest item with the trash can button, then got to keepass and copy a password.

**Comment by hluk** [comment]
This is how CopyQ gets source window title of the copied content (there is no better reliable solution I found): Window title is continuously checked whenever a current window changes (but this behavior depends a lot on window manager). If the window change is detected there is a small (~50ms) delay before any newly copied content is assigned a new window title.

This happens to work in most cases - mainly if your window manager hides right after a copy operation.

So in this case, it looks like CopyQ thinks that the source window is the previous one focused because password manager has focus for only a very short time when password is copied.

I've checked the content of clipboard when a password is copied from KeePassXC and there is now a special format `x-kde-passwordManagerHint`. With this we can use more reliable way to avoid storing passwords by checking if the format is present:

```ini
[Command]
Automatic=true
Command="
    copyq ignore"
Icon=\xf069
Input=x-kde-passwordManagerHint
Name=Ignore *\"Password\"* window
Remove=true
```

There is also [Blocklisted Texts](https://github.com/hluk/copyq-commands/tree/master/Scripts#blocklisted-texts) command which would show a quick notification allowing you to permanently block the newly copied content from CopyQ.

**Comment by bt103** [comment]
Ah, the x-kde-passwordManagerHint looks like it is the perfect reliable solution, thanks.  That solves it.

This last bit is not important, but if anybody knows...
I see there is a GUI for the commands, and the commands are also shown in raw form in these forums.  I think I've figured out how to insert and extract the raw commands into/out-of the GUI.  However I don't see a way to edit the commands once they are inserted.  You can edit the commands with the GUI, but it's a bit of feeling around in the dark, especially when the commands are shared in this forum in raw form.  Is there a way to directly edit the raw commands without exporting, editing, and re-importing?  I had a look through the doc and couldn't find an answer.

**Comment by hluk** [comment]
>  Is there a way to directly edit the raw commands without exporting, editing, and re-importing?

Editing the `ini` format directly in CopyQ is not supported. I still don't see a reason why do that. It is easier and safer to edit it in the Command dialog.
